quinn/src/connection.rs

```diff
@@ -742,71 +738,14 @@ impl io::Read for BiStream {
     }
 }
 
-impl AsyncRead for BiStream {
-    unsafe fn prepare_uninitialized_buffer(&self, _: &mut [u8]) -> bool {
-        false
-    }
-}
-
-/// A stream that can only be used to send data
-pub struct SendStream(BiStream);
-
-impl Write for SendStream {
-    fn poll_write(&mut self, buf: &[u8]) -> Poll<usize, WriteError> {
-        Write::poll_write(&mut self.0, buf)
-    }
-    fn poll_finish(&mut self) -> Poll<(), ConnectionError> {
-        self.0.poll_finish()
-    }
-    fn reset(&mut self, error_code: u16) {
-        self.0.reset(error_code);
-    }
-}
-
-impl io::Write for SendStream {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        self.0.write(buf)
-    }
-    fn flush(&mut self) -> io::Result<()> {
-        Ok(())
-    }
-}
-
-impl AsyncWrite for SendStream {
-    fn shutdown(&mut self) -> Poll<(), io::Error> {
-        self.0.shutdown()
-    }
-}
-
-/// A stream that can only be used to receive data
-pub struct RecvStream(BiStream);
-
-impl Read for RecvStream {
-    fn poll_read_unordered(&mut self) -> Poll<(Bytes, u64), ReadError> {
-        self.0.poll_read_unordered()
-    }
-    fn poll_read(&mut self, buf: &mut [u8]) -> Poll<usize, ReadError> {
-        Read::poll_read(&mut self.0, buf)
-    }
-    fn stop(&mut self, error_code: u16) {
-        self.0.stop(error_code)
-    }
-}
-
-impl io::Read for RecvStream {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        self.0.read(buf)
-    }
-}
-
 impl AsyncRead for RecvStream {
     unsafe fn prepare_uninitialized_buffer(&self, _: &mut [u8]) -> bool {
         false
     }
 }
 
 /// Uses unordered reads to be more efficient than using `AsyncRead` would allow
-pub fn read_to_end<T: Read>(stream: T, size_limit: usize) -> ReadToEnd<T> {
+pub fn read_to_end(stream: RecvStream, size_limit: usize) -> ReadToEnd {
```

Comment:
Would it make sense to move this to a method on `RecvStream`, now that it doesn't need to be generic?

Reply:
Yup.